source/common/config/api_type_oracle.cc

```diff
@@ -31,5 +31,20 @@ ApiTypeOracle::getEarlierVersionMessageTypeName(const std::string& message_type)
   }
   return absl::nullopt;
 }
+
+const absl::optional<std::string> ApiTypeOracle::getEarlierTypeUrl(const std::string& type_url) {
+  size_t type_name_start = type_url.find('/');
+  if (type_name_start == std::string::npos) {
+    return {};
+  }
+  std::string message_type = type_url.substr(type_name_start + 1);
```

Comment:
I forgot that there's also absl::StrSplit() which might be more readable than this, I'll leave it up to you.